#2711: Domain manager view - remove manager when only one manager [BOB]

dave-kennedy-ecs opened the pull request:
## Ticket

<!-- PR title format: `#issue_number: Descriptive name ideally matching ticket name - [sandbox]`-->
Resolves #2711 

## Changes

<!-- What was added, updated, or removed in this PR. -->
- Domain manager 'Remove' button is no longer disabled when there is only one domain manager
- When there is only one domain manager, and the 'Remove' button is clicked, an error message is returned, reading "Domains must have at least one domain manager."

<!--
    Please add/remove/edit any of the template below to fit the needs
    of this specific PR.
--->

## Context for reviewers

<!--Background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers.  -->
For developers reviewing, I added some notes in the PR where I removed or changed code and why.

## Setup

<!--  Add any steps or code to run in this section to help others run your code.
    
    Example 1:
    ```sh
    echo "Code goes here"
    ``` 
    
    Example 2: If the PR was to add a new link with a redirect, this section could simply be:
    -go to /path/to/start/page
    -click the blue link in the <insert location>
    -notice user is redirected to <proper end location>
-->

## Code Review Verification Steps

### As the original developer, I have

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Update documentation in READMEs and/or onboarding guide

#### Ensured code standards are met (Original Developer)
<!-- Mark "- N/A" and check at the end of each check that is not applicable to your PR -->
- [x] If any updated dependencies on Pipfile, also update dependencies in requirements.txt.
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [x] Tag @dotgov-designers in this PR's Reviewers for design review. If code is not user-facing, delete design reviewer checklist
- [x] Verify new pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

### As a code reviewer, I have

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Verified code meets all checks above. Address any checks that are not satisfied
- [ ] Reviewed this code and left comments. Indicate if comments must be addressed before code is merged
- [ ] Checked that all code is adequately covered by tests
- [ ] Verify migrations are valid and do not conflict with existing migrations

#### Validated user-facing changes as a developer
**Note:** Multiple code reviewers can share the checklists above, a second reviewer should not make a duplicate checklist. All checks should be checked before approving, even those labeled N/A. 

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### As a designer reviewer, I have

#### Verified that the changes match the design intention

- [x] Checked that the design translated visually
- [x] Checked behavior. Comment any found issues or broken flows.
- [x] Checked different states (empty, one, some, error)
- [x] Checked for landmarks, page heading structure, and links

#### Validated user-facing changes as a designer

- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Tested with multiple browsers (check off which ones were used)
  - [x] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [x] Safari
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### References
- [Code review best practices](../docs/dev-practices/code_review.md)

## Screenshots

<!-- If this PR makes visible interface changes, an image of the finished interface can help reviewers
and casual observers understand the context of the changes.
A before image is optional and can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow.
You may want to use [GIPHY Capture](https://giphy.com/apps/giphycapture) for this! 📸

_Please frame images to show useful context but also highlight the affected regions._
--->

dave-kennedy-ecs commented on src/registrar/tests/test_views_domain.py:
```diff
@@ -1317,6 +1318,57 @@ def test_domain_invitation_flow(self):
         self.assertContains(home_page, self.domain.name)
 
+    @less_console_noise_decorator
+    def test_domain_user_role_delete(self):
```
No previous tests existed for this view

dave-kennedy-ecs commented on src/registrar/tests/test_views_domain.py:
```diff
@@ -1258,8 +1259,8 @@ def test_domain_invitation_cancel_retrieved_invitation(self):
         self.assertContains(response, f"Invitation to {email_address} has already been retrieved.")
-        # Assert that the Cancel link is not displayed
-        self.assertNotContains(response, "Cancel")
+        # Assert that the Cancel link (form) is not displayed
+        self.assertNotContains(response, f"/invitation/{invitation.id}/cancel")
```
Prior test was not specific enough

dave-kennedy-ecs commented on src/registrar/views/utility/mixins.py:
```diff
@@ -343,12 +343,6 @@ def has_permission(self):
         if not (has_delete_permission or user_is_analyst_or_superuser):
             return False
 
-        # Check if more than one manager exists on the domain.
-        # If only one exists, prevent this from happening
-        has_multiple_managers = len(UserDomainRole.objects.filter(domain=domain_pk)) > 1
-        if not has_multiple_managers:
-            return False
-
```
This condition no longer returns a 403 permissions problem, and instead the view handles this and returns a proper error message.

dave-kennedy-ecs commented on src/registrar/views/domain.py:
```diff
@@ -1074,9 +1074,6 @@ def get_context_data(self, **kwargs):
         """The initial value for the form (which is a formset here)."""
         context = super().get_context_data(**kwargs)
 
-        # Add conditionals to the context (such as "can_delete_users")
-        context = self._add_booleans_to_context(context)
-
```
no longer needed in context since template is no longer disabling button based on context values this method was adding

dave-kennedy-ecs commented on src/registrar/templates/domain_users.html:
```diff
@@ -69,7 +69,6 @@ <h2 class> Domain managers </h2>
         </th>
         {% if not portfolio %}<td data-label="Role">{{ item.permission.role|title }}</td>{% endif %}
         <td>
-        {% if can_delete_users %}
```
conditional now obsolete

dave-kennedy-ecs commented on src/registrar/templates/domain_users.html:
```diff
@@ -112,18 +111,6 @@ <h2 class> Domain managers </h2>
               {% csrf_token %}
             </form>
           {% endif %}
-        {% else %}
-          <input 
-          type="submit" 
-          class="usa-button--unstyled disabled-button usa-tooltip usa-tooltip--registrar" 
-          value="Remove"
-          data-position="bottom"
-          title="Domains must have at least one domain manager"
-          data-tooltip="true"
-          aria-disabled="true"
-          role="button"
-          >
-        {% endif %}
```
no longer need to present disabled button under any conditions

gabydisarli commented:
Hi Dave! Can you change the screen reader output on the Remove button to include the email of the domain manager being removed like "Remove <email>"

dave-kennedy-ecs commented:
> Hi Dave! Can you change the screen reader output on the Remove button to include the email of the domain manager being removed like "Remove "

Yes. I set the aria-label to "Remove {{email}}" on the Remove link.

gabydisarli commented:
I'm just realizing I didn't quite update the alert message and it said the same thing as org model, which doesn't make sense. Can you update the message to say:

> Domains must have one domain manager. To remove yourself, the domain needs another domain manager.

Also, can you match the usual margin we have when alerts stack between the error alert that appears and the info alert below it

dave-kennedy-ecs commented:
> I'm just realizing I didn't quite update the alert message and it said the same thing as org model, which doesn't make sense. Can you update the message to say:
> 
> > Domains must have one domain manager. To remove yourself, the domain needs another domain manager.
> 
> Also, can you match the usual margin we have when alerts stack between the error alert that appears and the info alert below it

@gabydisarli I have updated the message. And I have modified the margin. Some of the alerts (across the app) have class margin-bottom-2 and some have margin-bottom-3. I updated the alerts on the domain (base) page to margin-bottom-2. I couldn't tell what was the 'usual' margin, so please test and verify that I did so correctly.

erinysong commented:
Got the expected alert when trying to remove myself as a domain's sole manager 
<img width="658" alt="image" src="https://github.com/user-attachments/assets/f3108d09-c914-48ec-a303-f393c30b9fa7" />

erinysong reviewed (approved):
LGTM - thank you for leaving comments on your PR, it helped a lot to more thoroughly understand what the diff was doing!

gabydisarli commented:
> > I'm just realizing I didn't quite update the alert message and it said the same thing as org model, which doesn't make sense. Can you update the message to say:
> > > Domains must have one domain manager. To remove yourself, the domain needs another domain manager.
> > 
> > 
> > Also, can you match the usual margin we have when alerts stack between the error alert that appears and the info alert below it
> 
> @gabydisarli I have updated the message. And I have modified the margin. Some of the alerts (across the app) have class margin-bottom-2 and some have margin-bottom-3. I updated the alerts on the domain (base) page to margin-bottom-2. I couldn't tell what was the 'usual' margin, so please test and verify that I did so correctly.

Oof. I'll add that to my list of code updates we need to make to unify our spacing which is a bit of a hodge podge of different patterns right now. `margin-bottom-2` is the correct spacing.
